src/packages/client/src/runtime/getPrismaClient.ts

```diff
@@ -595,6 +621,152 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
+      ;(params as InternalRequestParams).callsite = callsite
+
+      return this._executeRequest(params as InternalRequestParams)
+    }
+
+    private _executeRequest({
+      args,
+      clientMethod,
+      dataPath,
+      callsite,
+      runInTransaction,
+      action,
+      model,
+    }: InternalRequestParams) {
+      if (action !== 'executeRaw' && action !== 'queryRaw' && !model) {
+        throw new Error(`Model missing for action ${action}`)
+      }
+
+      if ((action === 'executeRaw' || action === 'queryRaw') && model) {
+        throw new Error(
+          `executeRaw and queryRaw can't be executed on a model basis. The model ${model} has been provided`,
+        )
+      }
+
+      let rootField: string | undefined
+      const operation = actionOperationMap[action]
+
+      if (action === 'executeRaw' || action === 'queryRaw') {
+        rootField = action
+      }
+
+      // TODO: Replace with lookup map for speedup
+      let mapping
+      if (model) {
+        mapping = this.dmmf.mappingsMap[model]
+        if (!mapping) {
+          throw new Error(
+            `Could not find mapping for model ${model} ${JSON.stringify(
+              arguments,
+            )}`,
+          )
+        }
+
+        rootField = mapping[action]
+      }
+
+      if (operation !== 'query' && operation !== 'mutation') {
+        throw new Error(`Invalid operation ${operation} for action ${action}`)
+      }
+
+      const field = this.dmmf.rootFieldMap[rootField!]
+
+      if (!field) {
+        throw new Error(
+          `Could not find rootField ${rootField} for action ${action} for model ${model} on rootType ${operation}`,
+        )
+      }
+
+      const { isList } = field.outputType
+      const typeName = getOutputTypeName(field.outputType.type)
+
+      let document = makeDocument({
+        dmmf: this.dmmf,
+        rootField: rootField!,
+        rootTypeName: operation,
+        select: args,
+      })
+
+      document.validate(args, false, clientMethod, this.errorFormat, callsite)
+
+      document = transformDocument(document)
+
+      // as printJsonWithErrors takes a bit of compute
+      // we only want to do it, if debug is enabled for 'prisma-client'
+      if (Debug.enabled('prisma-client')) {
+        const query = String(document)
+        debug(`Prisma Client call:`)
+        debug(
+          `prisma.${clientMethod}(${printJsonWithErrors({
+            ast: args,
+            keyPaths: [],
+            valuePaths: [],
+            missingItems: [],
+          })})`,
+        )
+        debug(`Generated request:`)
+        debug(query + '\n')
+      }
+
+      return this.fetcher.request({
+        document,
+        clientMethod,
+        typeName,
+        dataPath,
+        isList,
+        rootField: rootField!,
+        callsite,
+        showColors: this.errorFormat === 'pretty',
+        args,
+        engineHook: this._engineMiddlewares[0],
```

Comment:
Should engineHook be renamed engineMiddleware?

Reply:
The engines part is an internal api for now that we don't want to document.
I want to expose that properly once we need it.